helpers/helper_nano.py

```diff
@@ -282,6 +282,216 @@ def ZMuMu_MuSelection(df):
+
+
+def DQMOff_TauSelection(df):
+    '''
+    Selects Z->tautau events passing a single muon trigger. Defines probe tau pt/eta/phi
+    The selections are made to match with DQM Off selections
+    '''
+    df = df.Filter('HLT_IsoMu20')
+
+    # Trigged on a Muon (probably redundant)
+    df = df.Filter('''
+    bool trigged_on_mu = false;
+    for (unsigned int i = 0; i < TrigObj_id.size(); i++){
+        if(TrigObj_id[i] == 13) trigged_on_mu = true;
+    }
+    return trigged_on_mu;
+    ''')
+
+    # Charge 
+    # L1Mu_hwCharge = 0 corresponds to Muon_charge = +1
+    # L1Mu_hwCharge = 1 corresponds to Muon_charge = -1
+    # df = df.Define('L1Mu_charge', 'charge_conversion(L1Mu_hwCharge)')
+
+    # --------------------------------------------------------
+    # Match L1Mu to trig obj with id == 13 -> L1Mu_trig_idx
+    # with L1Mu_trig_idx -> check filter bits -> L1Mu_passHLT
+    # --------------------------------------------------------
+
+    # TrigObj matching
+    df = df.Define('Muon_trig_idx', 'MatchObjToTrig_DQMOff(Muon_eta, Muon_phi, TrigObj_pt, TrigObj_eta, TrigObj_phi, TrigObj_id, 13, TrigObj_filterBits, 3, 0.6)')
+    df = df.Define('Muon_passHLT_IsoMu24', 'trig_is_filterbit_set_DQMOff(Muon_trig_idx, TrigObj_filterBits, 3)')
+
+    df = df.Define('Muon_PassTightId','Muon_pfIsoId>=3&&Muon_mediumPromptId') 
+    df = df.Define('Muon_MassId', 'pass_muon_met_mass_below30(Muon_pt, Muon_phi, MET_pt, MET_phi)')
+
+    df = df.Define('isTag','Muon_pt>24&&abs(Muon_eta)<2.1&&abs(Muon_pdgId)==13&&Muon_PassTightId&&Muon_passHLT_IsoMu24&&Muon_MassId')
+    df = df.Filter('Sum(isTag)>0')
+
+    df = df.Define('tagIdx', 'first_tagmuon_idx(isTag)')
+    df = df.Filter('tagIdx!=-1')
+
+    df = df.Define('tagPt', 'Muon_pt[tagIdx]')
+    df = df.Define('tagEta', 'Muon_eta[tagIdx]')
+    df = df.Define('tagPhi', 'Muon_phi[tagIdx]')
+    df = df.Define('tagMass', 'Muon_mass[tagIdx]')
+    df = df.Define('tagCharge', 'Muon_charge[tagIdx]')
+
+    nEvts = df.Count().GetValue()
+    print("There are {} events after finding tag muon".format(nEvts))
+
+    # Decay mode 5 and 6 are unphysical, so removing them
+    # Applying Tau_Id_Discrimination against e, mu, jets
+    # Applying probe tau selection criterion with the given tag muon
+    df = df.Define('Tau_PassDecayMode', 'Tau_decayMode!=5&&Tau_decayMode!=6')
+    df = df.Define('Tau_PassTightId', 'Tau_idDeepTau2018v2p5VSe>=1&&Tau_idDeepTau2018v2p5VSjet>=2&&Tau_idDeepTau2018v2p5VSmu>=1')
+    df = df.Define('Tau_PassProbe', 'pass_probeTau(Tau_pt, Tau_eta, Tau_phi, Tau_mass, Tau_charge, tagPt, tagEta, tagPhi, tagMass, tagCharge)')
+
+    ## Commenting TauID cuts for testing. 
+    ## !! Need to uncomment later
+    df = df.Define('isProbeTau', 'Tau_PassDecayMode&&Tau_PassTightId&&Tau_PassProbe')
+    # df = df.Define('isProbeTau', 'Tau_PassProbe')
+    df = df.Filter('Sum(isProbeTau)>0')
+
+    nEvts = df.Count().GetValue()
+    print("There are {} events after finding probe Taus".format(nEvts))
+
```

Comment:
These two lines are triggering an event loop here. They should come later.

Reply:
Same as above. These lines were used for debugging/testing. Have been removed.